.gitlab-ci.yml

```diff
@@ -1,16 +1,21 @@
 stages:
   - unit-test
   - build
+  - deploy-dev
 
 run-unit-tests:
   stage: unit-test
+  # This resource group is configured with process_mode=oldest_first to make sure the pipelines are run serially.
```

Comment:
Is `oldest_first` process_mode default one? I do not see process_mode is set in this file?

Reply:
No, the default is `unordered`. We have to explicitly set process_mode to `oldest_first` using this mechanism - https://docs.gitlab.com/ee/api/resource_groups.html#edit-an-existing-resource-group. That is needed to make sure the pipeline is serial.

Reply:
Why do we need to run unit-test & build jobs in each pipeline serially? I don't see any issue if they run concurrently for different pipeline runs. It's only the deploy job and e2e tests that should run serially for each pipeline.

Reply:
There are a few issues with supporting concurrent pipelines with GitLab the way we want, so for now, I'm making our pipeline serial so that it works reliably. We can make the pipeline efficient in the future(task added to the backlog).

We can discuss offline the challenges in supporting concurrent pipelines in GitLab CI.